### PR TITLE
Migrate `remap-path-prefix`, `debug-assertions` and `emit-stack-sizes` `run-make` tests to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

Needs OSX/MSVC try jobs.

try-job: aarch64-apple
try-job: x86_64-msvc

### CODE REVIEW

**Comment by Oneirical** [comment]
The only change made was restoring the ignore-directives previously featured in the Makefile test, with added Reason comment.

@rustbot review

**Comment by Oneirical** [comment]
Backslash fun. Of course.

Instead of restoring the ignore statement, I think I can try to normalize the output.

File: tests/run-make/remap-path-prefix/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        .emit("metadata")
+        .input("auxiliary/lib.rs");
+    out_diagobj
+        .remap_path_prefix("auxiliary", "/the/aux")
+        .crate_type("lib")
+        .emit("metadata")
+        .input("auxiliary/lib.rs");
+
+    out_simple.run();
+    invalid_utf8_contains("liblib.rmeta", "/the/aux/lib.rs");
```
[the remark this answers is not in the document]
Comment (Oneirical) [reply tests/run-make/remap-path-prefix/rmake.rs:37]: That is true, it was supposed to be a one-off thing and it grew into something of a scale that shouldn't have been.

File: tests/run-make/remap-path-prefix/rmake.rs [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 // FIXME(Oneirical): check if works without ignore-windows
 
-use run_make_support::{invalid_utf8_contains, invalid_utf8_not_contains, is_darwin, rustc};
+use run_make_support::{fs_wrapper, is_darwin, rustc};
```
Comment (Kobzol) [inline tests/run-make/remap-path-prefix/rmake.rs:8]: I would still rename `invalid_utf8_contains` to e.g. `contains_bytes`, but doesn't need to be in this PR.

File: tests/run-make/remap-path-prefix/rmake.rs [hunk 3]
```diff
@@ -47,12 +47,51 @@ fn main() {
...
+//FIXME(Oneirical): These could be generalized into run_make_support
+// helper functions.
+fn rmeta_contains(expected: &str) {
+    // Normalize to account for path differences in Windows.
+    let haystack: Vec<u8> = fs_wrapper::read("liblib.rmeta")
+        .iter()
+        .map(|x| if *x == b'\\' { b'/' } else { *x })
+        .collect();
+    let needle = expected.as_bytes();
+    for i in 0..haystack.len() - needle.len() + 1 {
```
Comment (Kobzol) [inline tests/run-make/remap-path-prefix/rmake.rs:69]: This is hard to read and trust :/ What about something like this:
```rust
fn is_sub<T: PartialEq>(haystack: &[T], needle: &[T]) -> bool {
    haystack.windows(needle.len()).any(|c| c == needle)
}
```
(from https://stackoverflow.com/questions/47043167/does-rust-contain-a-way-to-directly-check-whether-or-not-one-vector-is-a-substr/47044053#47044053).

Or we can add https://docs.rs/bstr/latest/bstr/ to `run-make-support`.

**Comment by Kobzol** [comment]
Yeah, let's add `bstr` to `run-make-support`, it doesn't add any new dependencies. With `bstr`, the rmeta check function becomes something like `assert!(BString::from(slice).replace(b"\\", b"/").contains_str(slice))`.

**Comment by Oneirical** [comment]
> Yeah, let's add `bstr` to `run-make-support`, it doesn't add any new dependencies. With `bstr`, the rmeta check function becomes something like `assert!(BString::from(slice).replace(b"\\", b"/").contains_str(slice))`.

I just gave it a shot. Though this is a temporary solution, I'd like to make a new PR after this one that replaces all instances of `invalid_utf8_contains` with a `bstr` rework, with optional normalization.

**Comment by Kobzol** [comment]
Thanks! I assume that `remap-path-prefix` still fails on Windows?

@bors try

**Comment by Oneirical** [comment]
> Thanks! I assume that `remap-path-prefix` still fails on Windows?

I don't know, the normalization-backslash fix hasn't been tried yet. We're about to see.

File: tests/run-make/remap-path-prefix/rmake.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
+// Generating metadata alongside remap-path-prefix would fail to actually remap the path
+// in the metadata. After this was fixed in #85344, this test checks that "auxiliary" is being
+// successfully remapped to "/the/aux" in the rmeta files.
+// See https://github.com/rust-lang/rust/pull/85344
+
+// FIXME(Oneirical): check if works without ignore-windows
```
Comment (Kobzol) [inline tests/run-make/remap-path-prefix/rmake.rs:6]: You can remove fixme, looks like it works :)

**Review by Kobzol — APPROVED** [review]

File: tests/run-make/remap-path-prefix/rmake.rs (hunk 4 above)
Comment (Oneirical) [reply tests/run-make/remap-path-prefix/rmake.rs]: Done!

**Comment by jieyouxu** [comment]
r? Kobzol

**Pull request merged**